http-webhook.md

```diff
@@ -0,0 +1,339 @@
+observe the value of the Retry-After header and refrain from sending further
+requests until the indicated time.
+
+If the delivery has been accepted and processed, and if the response carries a
+payload with processing details, the response MUST have the [200 OK][200] or
+[201 Created][201] status code. In this case, the response MUST carry a
+[`Content-Type`][Content-Type] header.
+
+If the delivery has been accepted and processed, but carries no payload, the
+response MUST have the [201 Created][201] or [204 No Content][204] status code.
+
+If the delivery has been accepted, but has not yet been processed or if the
+processing status is unknown, the response MUST have the [202 Accepted][202]
+status code.
+
+If the delivery cannot be accepted because the notification format has not
+been understood, the service MUST respond with status code [415][415].
+
+All further error status codes apply as specified in [RFC7231][RFC7231].
+
+## 3. Authorization
+
+The delivery request MUST use one of the following two methods, both of which
+lean on the OAuth 2.0 Bearer Token [RFC6750][RFC6750] model.
+
+The delivery target MUST support both methods.
+
+The client MAY use any token-based authorization scheme. Challenge-based
+schemes MUST NOT be used.
+
+### 3.1.  Authorization Request Header Field
+
+The access token is sent in the [`Authorization`][Authorization] request
+header field defined by HTTP/1.1.
+
+For [OAuth 2.0 Bearer][Bearer] tokens, the "Bearer" scheme MUST be used.
+
+Example:
+
+``` text
+POST /resource HTTP/1.1
+Host: server.example.com
+Authorization: Bearer mF_9.B5f-4.1JqM
+```
+
+### 3.2 URI Query Parameter
+
+When sending the access token in the HTTP request URI, the client adds the
+access token to the request URI query component as defined by "Uniform Resource
+Identifier (URI): Generic Syntax" [RFC3986][RFC3986], using the "access_token"
+parameter.
+
+For example, the client makes the following HTTP request:
+
+``` text
+POST /resource?access_token=mF_9.B5f-4.1JqM HTTP/1.1
+Host: server.example.com
+```
+
+The HTTP request URI query can include other request-specific parameters, in
+which case the "access_token" parameter MUST be properly separated from the
+request-specific parameters using "&" character(s) (ASCII code 38).
+
+For example:
+
+    https://server.example.com/resource?access_token=mF_9.B5f-4.1JqM&p=q
+
+Clients using the URI Query Parameter method SHOULD also send a
+Cache-Control header containing the "no-store" option.  Server
+success (2XX status) responses to these requests SHOULD contain a
+Cache-Control header with the "private" option.
+
+Because of the security weaknesses associated with the URI method (see
+[RFC6750, Section 5][RFC6750]), including the high likelihood that the URL
+containing the access token will be logged, it SHOULD NOT be used unless it is
+impossible to transport the access token in the "Authorization" request header
```

Comment:
There are plenty of sites where you can configure web hooks that will allow you to paste a URL and nothing else. Having a token right in the URL is simply the easiest way to give someone a reference with embedded access permission. So it's not neccessarily about not having access to the header per-se (which still is an occasional problem w/ some APIs), but about the information being split across two places. The feature is already optional for the client, since that's where those concerns arise. The server shouldn't have a opt-out choice given that it is a client consideration and might be a constraint for which the client has no alternative.

Reply:
> The feature is already optional for the client, since that's where those concerns arise.

I don't agree it is only a concern for the client, and it's been an issue for me several times (e.g. with IronMQ urls, and with Azure Functions - both default to tokens in the URI).

Our clients entrust us with their access tokens, and they should be assured we keep them safe (similar to e.g. passwords, or personal data of our clients customers).
We have policies to never log passwords, or personal data, or similar. Access to our logs is not heavily restricted, allowing many people within our company to quickly gain insights if things go wrong (e.g. support).
If my code logs the access token, I am a) basically violating company policy if you assume access tokens are a form of a password, and b) make the access token available to a broad range of people within my company, that are not allowed to access critical data.

(As a commerce system, stealing an access token may allow the attacker to e.g. trigger a refund, or trigger the shipment of an order before the payment is confirmed - so the threat isn't purely theoretical)

Fair enough, then don't log the URI, right? Turns out it is virtually impossible to assure if you are using an HTTP lib on the JVM (I'll spare you the details...).

Reply:
> Having a token right in the URL is simply the easiest way to give someone a reference with embedded access permission.

I totally agree.

Here is a compromise: The delivery target is allowed to register the webhook with the access token in the URI. The sender MUST send the access token, but it SHOULD use the Authorization Request Header Field instead of the URI Query Parameter.

(I guess this is technically already fine via `The delivery target MUST support both methods.`, but it would be surprising for the client)

Reply:
@cneijenhuis In our service, we have "Connection Strings" that have URLs and AuthNZ material in them, and we then split that information across query string and headers as needed. 

I *like* the rule you stated above, but it's right on the line of getting into the hair of how an implementation does its config. We don't mention the actual subscription gesture in any spec (and per earlier group discussions it'll be quite difficult to harmonize it), and therefore it's difficult to impose constraints on it. 

I already make a pretty strong and normative SHOULD NOT statement against the URI method, including giving caching guidance, and that effectively says what you say here. How about we amend that with:

"Even if the client has been configured with a token embedded in the URI query string conforming with 3.2, the client SHOULD extract and remove the token from the query string and send it via the Authorization request header field whenever feasible."

Reply:
@clemensv Sounds great 👍 

Minor: `the client SHOULD extract` should probably be `the sender SHOULD extract`?